system/security/pkencrypt/pkencrypt.cpp

```diff
@@ -0,0 +1,260 @@
+#ifdef _USE_ZLIB
+        CriticalBlock b(decrypterCrit);
+
+        if (!pDecrypter)
+        {
+            StringBuffer privateKeyBuff;
+            privateKeyBuff.loadFile(privateKeyFile.str());
+            if (privateKeyBuff.isEmpty())
+                throw MakeStringException(-1, "PKEncrypt error: Could not read private key file");
+            pDecrypter = createZDecryptor(privateKeyBuff.str(), nullptr);//create once
+            if (pDecrypter)
+            {
+                pDecrypter->setEncoding(true);
+#ifdef _DEBUG
+                pDecrypter->setTraceLevel(100);
+#endif
+            }
+            else
+                throw MakeStringException(-1, "PKEncrypt error: could not create IZDecryptor");
+        }
+
+        unsigned char * pKey = (unsigned char *)_key;
+        unsigned char * pToken = (unsigned char *)_token;
+        IZBuffer* result = pDecrypter->decrypt(pKey, pToken);
+        if (result)
+        {
+            _result.set((char*)result->buffer(), result->length());
+            releaseIZ(result);
+            return true;  //success
+        }
+        else
+            throw MakeStringException(-1, "PKEncrypt error: decrypter return NULL");
+#else
+        WARNLOG("PKEncrypt error:: Platform built without ZLIB!");
+#endif
+        return false;//failure
+    }
+
+
+    //encrypt workunit security token
+    //Format is WUID;USER;PASSWORD
+    virtual bool createWUSecurityToken(const char * _wuid, const char * _user, const char * _password, IStringVal & _key, IStringVal & _token)
+    {
+        if (!encrpytConfigured)
+            throw MakeStringException(-1, "PKEncrypt error: encryption not enabled");
+#ifdef _USE_ZLIB
+        VStringBuffer str("%s;%s;%s", _wuid, _user, _password);
+        StringAttr tokStr(str);
+        StringAttr key;
+        StringAttr token;
+        bool rc = PKEncrypt(str, key, token);
+        if (rc)
+        {
+            _key.set(key);
+            _token.set(token);
+            return true;//success
+        }
+        else
+            throw MakeStringException(-1, "PKEncrypt error: %d PKI encrypting token using key file %s", rc, publicKeyFile.str());
+#else
+        throw MakeStringException(-1, "PKEncrypt error: Platform built without ZLIB!");
+#endif
+        return false;//fail
+    }
+
+    //decrypt and tokenize workunit security token
+    //Format is WUID;USER;PASSWORD
+    virtual bool extractWUSecurityToken(const char * _key, const char * _token, const char * _wuid, IStringVal & _user, IStringVal & _password)
+    {
+        if (!decrpytConfigured)
+            throw MakeStringException(-1, "PKEncrypt error: decryption not configured");
+#ifdef _USE_ZLIB
+        StringAttr str;
+        bool ok = PKDecrypt(_key, _token, str);
+        if (ok)
+        {
+            char * pBuffer = (char*)str.str();
+            const char * finger = strchr(pBuffer, ';');
+            bool wuidsMatch = _wuid ? (finger && (0 == strnicmp(_wuid, pBuffer, finger - pBuffer))) : true;
```

Comment:
How often would this be called with wuid NULL / empty ?
If never then perhaps we could use the wuid itself in a simple encryption scheme here ?